keps/sig-storage/2268-non-graceful-shutdown/README.md

```diff
@@ -0,0 +1,469 @@
+
+* After 300 seconds (default), the Taint Manager tries to delete Pods on the Node after detecting that the Node is NotReady. The Pods will be stuck in terminating status.
+
+Proposed logic change:
+* The Pod GC Controller, part of the kube-controller-manager, would need to go through all the Pods in terminating state, verify that the Node is NotReady, check if the `Reason` in the `PodStatus` is set to the `NodeShutDownReason` constant, and then check if the `SafeDetach` flag in CSIDriver is set to true. If so, forcefully delete pods that are stuck.
+
+* Specifically the Pod GC Controller (the only controller able to forcefully delete pods) would also check now for stateful pods, and select those that satisfy ALL of the following conditions:
+
+  * pods that are backed by a CSI driver that supports `SafeDetach`.
+  * pods that are backed by volumes that aren't marked as `ReadWriteMany`.
+
+* Once pods are selected and forcefully deleted, the attachdetach reconciler should check if the `SafeDetach` flag in CSIDriver is set to true. If so, it should skip the `attachedVolume.MountedByNode` check for the volumes backing these pods and allow `volumeAttachement` to be deleted.
+
+* This would trigger the deletion of the `volumeAttachement` objects. This would allow `ControllerUnpublishVolume` to happen before `NodeUnpublishVolume` and/or `NodeUnstageVolume` are called.
+
+* When the `external-attacher` detects the `volumeAttachement` object is being deleted, it calls CSI driver's `ControllerUnpublishVolume`. The CSI driver would need to ensure that no volume is being used through a mount point. If unable to determine the usage, the CSI driver can return an error without detaching the volume. The `external-attacher` would then retry with backoffs.
+
+* The Pod GC Controller would also apply a `quarantine` taint on the nodes whose pods are forcefully deleted due to the shutdown. This should happen before the pods are being evicted. The "quarantine" taint indicates that new pods should not be scheduled on the node.
+
+### Handle the Return of Shutdown Node
+
+If the nodes that were previously shutdown with pods forcefully deleted come back, we need to make sure they are cleaned up before pods are scheduled on them again.
+
+The health check in Node lifecycle controller, part of kube-controller-manager, sets Node v1.NodeReady Condition to True if lease is renewed again. Node Status becomes Ready.
+
+When the node comes up, Kubelet volume manager reconciler will try to clean up mounts by calling `NodeUnpublishVolume`. Kubelet volume manager reconciler will also try to reconstruct global mount for CSI volumes and call `NodeUnstageVolume`. If there isn't enough information to reconstruct global mount, Kubelet volume manager should cleanup the [global mount directory](/var/lib/kubelet/plugins/kubernetes.io/csi/pv/{pvname}/globalmount) to avoid problems when new pods are scheduled to this node again.
```

Comment:
When volume reconstruction runs, is it guaranteed that kubelet pod manager is updated with the latest API state? I remember kubelet has implemented checkpointing to be able to recover state after a restart.

Reply:
If reconstruction is not able to cleanup properly, what should we do?

Reply:
It looks like sig-node didn't want to go with Kubelet checkpoint.  See comments here:
https://github.com/kubernetes/kubernetes/issues/489#issuecomment-868768740
The Kubelet checkpoint-restore KEP has been updated to address a different issue - pod migration between nodes:
https://github.com/kubernetes/enhancements/pull/1990